crmsh/bootstrap.py

```diff
@@ -331,34 +337,65 @@ def _update_configuration(self):
         Update /etc/sysconfig/sbd
         """
         shutil.copyfile(self.SYSCONFIG_SBD_TEMPLATE, SYSCONFIG_SBD)
+        self._determine_sbd_watchdog_timeout()
         sbd_config_dict = {
                 "SBD_PACEMAKER": "yes",
                 "SBD_STARTMODE": "always",
                 "SBD_DELAY_START": "no",
                 "SBD_WATCHDOG_DEV": self._watchdog_inst.watchdog_device_name
                 }
+        if self._sbd_watchdog_timeout > 0:
+            sbd_config_dict["SBD_WATCHDOG_TIMEOUT"] = str(self._sbd_watchdog_timeout)
         if self._sbd_devices:
             sbd_config_dict["SBD_DEVICE"] = ';'.join(self._sbd_devices)
         utils.sysconfig_set(SYSCONFIG_SBD, **sbd_config_dict)
         csync2_update(SYSCONFIG_SBD)
 
-    def _determine_stonith_watchdog_timeout(self):
+    def _determine_sbd_watchdog_timeout(self):
+        """
+        When using diskless SBD, determine value of SBD_WATCHDOG_TIMEOUT
+        """
+        if not self.diskless_sbd:
+            return
+        # add sbd after qdevice started
+        if utils.is_qdevice_configured() and utils.service_is_active("corosync-qdevice.service"):
+            qdevice_sync_timeout = utils.get_qdevice_sync_timeout()
+            self._sbd_watchdog_timeout = qdevice_sync_timeout + 5
+            if self._is_s390 and self._sbd_watchdog_timeout < 15:
+                self._sbd_watchdog_timeout = 15
+            self._stonith_timeout = self.calculate_stonith_timeout(self._sbd_watchdog_timeout, check_exist=True)
+        # add sbd and qdevice together from beginning
+        elif _context.qdevice_inst:
+            self._sbd_watchdog_timeout = self.SBD_WATCHDOG_TIMEOUT_DEFAULT
+            self._stonith_timeout = self.calculate_stonith_timeout(self._sbd_watchdog_timeout)
+
+    @staticmethod
+    def calculate_stonith_watchdog_timeout(sbd_watchdog_timeout, is_s390=False, check_exist=False):
         """
         Determine value of stonith-watchdog-timeout
         """
-        conf = utils.parse_sysconfig(SYSCONFIG_SBD)
-        res = conf.get("SBD_WATCHDOG_TIMEOUT")
-        if res:
-            self._stonith_watchdog_timeout = -1
-        elif "390" in os.uname().machine:
-            self._stonith_watchdog_timeout = "30s"
+        if sbd_watchdog_timeout:
+            if check_exist:
+                exist_swt = cib_factory.get_property("stonith-watchdog-timeout")
+                if not exist_swt:
+                    return -1
+                stonith_watchdog_timeout = int(exist_swt.strip('s'))
+                if stonith_watchdog_timeout >= (int(sbd_watchdog_timeout) * 2):
+                    return "{}s".format(stonith_watchdog_timeout)
+                else:
+                    return -1
+            else:
+                return -1
+        elif is_s390:
+            return "30s"
```

Comment:
AFAICT, it would be more straightforward to move this into the initial definition of `STONITH_WATCHDOG_TIMEOUT_DEFAULT`

Reply:
Changed